### ISSUE TITLE
Cannot easily change warnings in type declaration

### ISSUE THREAD

**Issue by goldfirere** [ocaml/ocaml#12301]
If I say

```ocaml
type t = A [@deprecated "blah"]

type t2 = t = A
```

then, with warning 3 enabled, I get a warning. Good. But suppose I want to locally suppress this warning. Neither of the following work:

```ocaml
type t2 = t = A [@warning "-3"]
type t2 = t = A [@@warning "-3"]
```

On the other hand, preceding `t2` with `[@@@warning "-3"]` works fine.

I think one of the more local forms should work.

**Comment by gasche** [comment]
That seems very reasonable to me. But then I am not sure, as a user, how I would reason about the situation where a given constructor and its re-export use different warning settings. When using `A`, when can I expect the behavior specified by `t` and the behavior specified by `t2`? Does the implementation coincide with my expectations?

My best guess at a specification that users could expect would be as follows:
- if I use explicit qualification `M.A`, and this unambiguously denotes only one of the declarations (because the other is is another module), then I get that constructor-declaration's warning behavior
- if I use type-directed disambiguation for just `A`, then I would consider that "which `A` I get" is not well-defined as long as both declarations are visible in my environment (but: the type-checker will do something, and users will rely on this behavior, which may get us into trouble if the type-checker behavior changes later; avoiding the situation where a re-export uses different warnings would remove this risk)
- when only one version is in my typing environment, then obviously I only see this version; this could be a not-uncommon case with `dune` aggressively hiding transitive dependencies (but it may change if the build system starts relying on #11989 instead).

Actually, what happens today? Do re-exports implicitly get the same warning behaviors as the original declaration? (I would guess: they don't.) Is the observed behavior reliable enough in practice?

This is related to #11867 which is at the level of module aliases (instead of re-export of type constructors).

**Comment by goldfirere** [comment]
I think these are good questions. But I propose that they are distinct from this ticket -- I intend this ticket to be about warnings that do or don't get printed at the re-export site, whereas @gasche's questions above are all about downstream usage sites. I don't expect any change to the warning behavior at the re-export site to have an impact on warnings printed at usage sites.

In any case, just to align expectations: I'm happy to tackle this, but I will be on holiday for the next several weeks, so don't expect any action until July.

**Comment by gaganchandan** [comment]
I am interested in attempting to tackle this. It would be great if someone could point to towards the relevant parts of the code base and let me know how I could get started.

**Comment by goldfirere** [comment]
The key function is `Builtin_attributes.warning_scope`, which changes the set of enabled warnings from a `[@warning "xyz"]` attribute (or with more `@` signs -- the number of `@` signs affects parsing, but all forms of attribute look identical from within the compiler). I see a call to this function from within `transl_declaration` in `transl_type_decl` in `typing/typedecl.ml`. But somehow it's not working: there must be some other code within `transl_type_decl` that processes the `t` or the `A` (in the example at the top of this ticket) outside of `transl_declaration`. So the first step would be to find the spot that actually emits the warning. My guess is that it's outside of this call to `warning_scope`. Then, move the call to `warning_scope` as appropriate.

What's slightly challenging here is that a natural spot for `warning_scope` is at the top of `transl_type_decl`, but that function processes an entire potentially-mutually-recursive group of type definitions, each of which might have separate `[@@warning "xyz"]` attributes. So we might need calls to `warning_scope` at each place within `transl_type_decl` that iterates through the list of decls. But before doing this, the first step is to figure out what's going wrong in the concrete example in this ticket, I think.

Does this have enough to get you moving? Thanks @gaganchandan for volunteering here!

**Comment by gaganchandan** [comment]
@goldfirere I think it should be enough for getting started. I'll look the through the parts mentioned and ask here if I have further questions. Thanks!

**Comment by gaganchandan** [comment]
I recently got some time to go through the code (sorry for the delay, I've been a little busy with my semester starting). The warning seems to be emitted inside the call to `check_abbrev` which is called during an iteration through the list of decls inside `transl_type_decl`. I tried moving the call into `warning_scope` but it doesn't seem to work:

```ocaml
  let check_abbrev env name_sdecl (id, slot) =
    Builtin_attributes.warning_scope name_sdecl.ptype_attributes (fun () ->
        check_abbrev env name_sdecl (id, slot))
  in
  List.iter2 (check_abbrev final_env) sdecl_list decls;
```

@goldfirere what do you think?

**Comment by goldfirere** [comment]
That looks pretty plausible to me. Did you try debugging with that change in force? Maybe look at the contents of `ptype_attributes` there to make sure nothing has been lost? For what it's worth, I would expect this to work with the `[@@warning "-3]`, but not with the `[@warning "-3"]`.

**Comment by gaganchandan** [comment]
You're right, it works with `[@@warning "-3"]`. I'll try to debug by looking at `ptype_attributes`.

**Comment by gaganchandan** [comment]
With a single `@`, `ptype_attributes` seems to be empty which is probably why it only works with two `@`s.

**Comment by gaganchandan** [comment]
@goldfirere how should I go about making it work with a single `@`, seeing as it doesn't seem to work with `warning_scope`? Or is it enough if one of the forms work?

**Comment by goldfirere** [comment]
It's definitely enough for just `[@@warning]`. The difference in the two is parsing precedence. When I say

```
type t2 = t = A [@warning "-3"]
```

that's like saying

```
type t2 = t = (A [@warning "-3"])
```

That is, the attribute ends up on the constructor (if it ends up anywhere at all). Single-`@` attributes attach themselves to a small bit of syntax.

On the other hand, writing

```
type t2 = t = A [@@warning "-3"]
```

is like

```
(type t2 = t = A) [@@warning "-3"]
```

A double-`@` attribute attaches itself to an entire declaration in a `struct` or `sig` -- pretty much what we want here.

(A triple-`@` attribute is an entire declaration in a `struct` or `sig`, unattached to any other syntax.)

So I'd expect only the double-`@` form to work, because it actually includes the usage of the deprecated `t`. My initial report tried both just for thoroughness.

Perhaps confusingly, OCaml also often accepts

```
type[@blah] t = A
```

as a shorthand for

```
type t = A [@@blah]
```

Thus, I'd expect your patch also to work for

```
type[@warning "-3"] t2 = t = A
```

without any changes.

Bottom line: it sounds like you're in good shape here! Ready to post a PR?

**Comment by gaganchandan** [comment]
Thanks for the clarification. @goldfirere. I think I can go ahead with a PR. Should I add tests too?

**Comment by goldfirere** [comment]
Yes, please. Thank you!
